### PR TITLE
[ONNX] Add repro export from `GraphInfo`

### PR DESCRIPTION
**Pull request by BowenBao**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #89947
* #89946
* #89808
* #89807

### CODE REVIEW

File: test/onnx/test_verification.py [hunk 1]
```diff
@@ -258,6 +259,18 @@ def forward(self, x):
             )
         self.assertExpected(f.getvalue())
 
+    def test_export_repro_for_mismatch(self):
+        mismatch_leaves = self.graph_info.all_mismatch_leaf_graph_info()
+        self.assertTrue(len(mismatch_leaves) > 0)
+        leaf_info = mismatch_leaves[0]
+        temp_dir = tempfile.TemporaryDirectory()
```
Comment (justinchuby) [inline test/onnx/test_verification.py:266]: Use it as a context manager so the directory is cleaned up?

File: torch/onnx/_internal/onnx_proto_utils.py [hunk 2]
```diff
@@ -12,6 +14,98 @@
 from torch.onnx._internal import _beartype, jit_utils, registration
 
 
+@_beartype.beartype
+def export_as_test_case(
+    model_bytes: bytes, inputs_data, outputs_data, name: str, dir: str
+):
```
Comment (justinchuby) [inline torch/onnx/_internal/onnx_proto_utils.py:20]: return type

File: torch/onnx/_internal/onnx_proto_utils.py [hunk 3]
```diff
@@ -12,6 +14,98 @@
...
+    try:
+        import onnx
+    except ImportError:
+        raise ImportError(
+            "Export test case to ONNX format failed: Please install ONNX."
+        )
+
+    test_case_dir = os.path.join(dir, "test_" + name)
```
Comment (justinchuby) [inline torch/onnx/_internal/onnx_proto_utils.py:28]: Do we need to ensure uniqueness?

File: torch/onnx/_internal/onnx_proto_utils.py [hunk 4]
```diff
@@ -12,6 +14,98 @@
...
+    output_files = glob.glob(os.path.join(test_data_dir, "output_*.pb"))
+    for output_file in output_files:
+        tensor = onnx.load_tensor(output_file)
+        outputs[tensor.name] = numpy_helper.to_array(tensor)
+
+    return model_bytes, inputs, outputs
+
+
+@_beartype.beartype
+def export_data(data, value_info_proto, f: str) -> None:
```
Comment (justinchuby) [inline torch/onnx/_internal/onnx_proto_utils.py:81]: Helpful to add a brief summary for all new functions?

File: torch/onnx/verification.py [hunk 5]
```diff
@@ -1070,25 +1105,82 @@ def pretty_print(self):
...
+            inputs: Inputs to the model.
+            outputs: Outputs of the model.
+            dir: Directory to save the repro.
+            name: Name of the test case. If not specified, a name based on current time
+                will be generated.
+        Returns:
+            Path to the repro.
+        """
+        if name is None:
+            name = datetime.datetime.now().strftime("%Y_%m_%d_%H_%M_%S_%f")
```
Comment (justinchuby) [inline torch/onnx/verification.py:1157]: Add a random string after the datetime maybe?

File: torch/onnx/verification.py [hunk 6]
```diff
@@ -1070,25 +1105,82 @@ def pretty_print(self):
...
-                node_kind = node_kinds.pop()
-                mismatch_node_kinds[node_kind] = (
-                    mismatch_node_kinds.get(node_kind, 0) + 1
-                )
-        print(" Mismatch node kinds: ".center(80, "="))
-        print(mismatch_node_kinds)
+
+    @_beartype.beartype
+    @classmethod
+    def create_test_case_repro(
```
Comment (justinchuby) [inline torch/onnx/verification.py:1141]: @_beartype.beartype may need to be closest to the def

**Review by justinchuby — APPROVED** [review]
LGTM🚀

File: torch/onnx/_internal/onnx_proto_utils.py (hunk 3 above)
Comment (BowenBao) [reply torch/onnx/_internal/onnx_proto_utils.py:28]: Good catch, very likely. I first written as using datetime but later extended to user provided name. Let me check.

File: torch/onnx/verification.py (hunk 5 above)
Comment (BowenBao) [reply torch/onnx/verification.py:1157]: 😄 This is getting to microseconds .. Should be safe enough?

File: torch/onnx/_internal/onnx_proto_utils.py (hunk 3 above)
Comment (BowenBao) [reply torch/onnx/_internal/onnx_proto_utils.py:28]: I think it's ok, it overrides if exists. This only happens when name is provided by user.
If user doesn't provide the name, the path can be considered always unique.

**Pull request closed without merging**